libraries/protocol/include/scorum/protocol/betting/market.hpp

```diff
@@ -2,16 +2,188 @@
 #include <scorum/protocol/betting/wincase.hpp>
 #include <scorum/protocol/betting/wincase_comparison.hpp>
 
+#include <boost/fusion/container.hpp>
+#include <boost/fusion/algorithm/iteration/for_each.hpp>
+#include <boost/fusion/include/for_each.hpp>
+#include <boost/container/flat_set.hpp>
+
+#include <scorum/utils/static_variant_serialization.hpp>
+
 namespace scorum {
 namespace protocol {
 namespace betting {
-struct market_type
+
+/// This check if strict_wincase_pair_type has valid market kind
+template <typename WincasePair> constexpr bool check_market_kind(market_kind kind, WincasePair&& w)
+{
+    // Only one check for left_wincase is necessary because strict_wincase_pair_type type
+    // guarantees that left_wincase and right_wincase have the same market kind
+    return kind == std::decay_t<decltype(w)>::left_wincase::kind_v;
```

Comment:
If you will use const lvalue reference then you won't need to use std::decay_t here